ql/experimental/asian/analytic_discr_geom_av_price_heston.hpp

```diff
@@ -0,0 +1,137 @@
+#include <ql/exercise.hpp>
+
+namespace QuantLib {
+
+    //! Pricing engine for European discrete geometric average price Asian
+    /*! This class implements a discrete geometric average price
+        Asian option with European exercise under the Heston stochastic
+        vol model where spot and variance follow the processes
+
+        \f[
+        \begin{array}{rcl}
+        dS(t, S)  &=& (r-d) S dt +\sqrt{v} S dW_1 \\
+        dv(t, S)  &=& \kappa (\theta - v) dt + \sigma \sqrt{v} dW_2 \\
+        dW_1 dW_2 &=& \rho dt \\
+        \end{array}
+        \f]
+
+        References:
+
+        Implements the analytical solution for continuous geometric Asian
+        options developed in "A Recursive Method for Discretely Monitored
+        Geometric Asian Option Prices", B. Kim, J. Kim, J. Kim & I. S. Wee,
+        Bull. Korean Math. Soc. 53, 733-749 (2016)
+
+        \ingroup asianengines
+
+        \test
+        - the correctness of the returned value is tested by reproducing
+              results in Tables 1, 2 and 3 of the paper
+
+        \todo handle seasoned options
+    */
+    class AnalyticDiscreteGeometricAveragePriceAsianHestonEngine
+        : public DiscreteAveragingAsianOption::engine {
+      public:
+        explicit AnalyticDiscreteGeometricAveragePriceAsianHestonEngine(
+            const ext::shared_ptr<HestonProcess>& process,
+            Real xiRightLimit = 100.0);
+        void calculate() const;
+
+        // Equation (21) - must be public so the integrand can access it.
+        std::complex<Real> Phi(std::complex<Real> s,
+                               std::complex<Real> w,
+                               Time t, Time T, Size kStar,
+                               const std::vector<Time>& t_n,
+                               const std::vector<Time>& tauK) const;
+
+      private:
+        // Initial process params
+        Real v0_, rho_, kappa_, theta_, sigma_, logS0_;
+        Handle<YieldTermStructure> dividendYield_;
+        Handle<YieldTermStructure> riskFreeRate_;
+        Handle<Quote> s0_;
+
+        ext::shared_ptr<HestonProcess> process_;
+
+        // A lookup table for the reuslts of omega_tilde() to avoid repeated calls for given Phi call
+        mutable std::map<Size, std::complex<Real> > omegaTildeLookupTable_;
+
+        // Cutoff parameter for integral in Eqs (23) and (24)
+        Real xiRightLimit_;
+
+        // Integrator for equation (23) and (24)
+        GaussLegendreIntegration integrator_;
+
+        // We need to set up several variables inside calculate as they depend on fixing times. Rather
+        // than pass them between a, omega, F etc. which makes for very messy method signatures, we
+        // make them mutable class properties instead.
+        mutable Real tr_t_;
+        mutable Real Tr_T_;
+        mutable std::vector<Real> tkr_tk_;
+
+        // Equation (11)
+        std::complex<Real> F(std::complex<Real>& z1,
+                             std::complex<Real>& z2,
```

Comment:
These would be either `const std::complex<Real>&` or `std::complex<Real>` if the inputs are not modified.  The copy is probably cheap.  Same below.

Reply:
Fixed throughout for both the discrete and the geometric pricer - no copying required now (these are passed through to each of the various functions).